### PR TITLE
Qualcomm AI Engine Direct - Support topk

### PR DESCRIPTION
**Pull request by winskuo-quic**
## Summary
- Support topK
- Properly decompose einsum for quantization annotation to work properly
- Unify warning messages in op builder
- Add UT

### CODE REVIEW

**Comment by winskuo-quic** [comment]
Hi @cccclai,
This PR is to enable topK op builder and properly decompose einsum for quantization.
Please have a look.
Thanks.

File: backends/qualcomm/passes/decompose_einsum.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+from executorch.exir.pass_base import ExportPass, PassResult
+from torch.fx.experimental.proxy_tensor import make_fx
+
+
+class DecomposeEinsum(ExportPass):
```
Comment (cccclai) [inline backends/qualcomm/passes/decompose_einsum.py:12]: This pattern seems common enough and we will see if we support this kind of logic more easily.

File: backends/qualcomm/tests/models.py [hunk 2]
```diff
@@ -416,6 +416,17 @@ def forward(self, x):
         return torch.sum(self.first(x), dim=(2, 3), keepdim=False)
 
 
+class Conv2dTopK(torch.nn.Module):
```
Comment (cccclai) [inline backends/qualcomm/tests/models.py:419]: I assume it's just to test this kind of common pattern, and they won't be ussed?

File: backends/qualcomm/tests/models.py [hunk 3]
```diff
@@ -440,6 +451,30 @@ def forward(self, x):
...
+    def forward(self, bn, anm, bm):
+        return torch.einsum("bn,anm,bm->ba", bn, anm, bm)
+
+
+class EinsumOuterProduct(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, i, j):
+        return torch.einsum("i,j->ij", i, j)
```
Comment (cccclai) [inline backends/qualcomm/tests/models.py:467]: Is einsum used somewhere?

**Review by cccclai — APPROVED** [review]
This looks great! In the long term, it will be great to seperate changes such that one PR is for one feature. This PR seems including other stuff more than `topk`

File: backends/qualcomm/tests/models.py (hunk 2 above)
Comment (winskuo-quic) [reply backends/qualcomm/tests/models.py:419]: Yes. This is just for testing purpose to check layout transform is working properly.

File: backends/qualcomm/tests/models.py (hunk 3 above)
Comment (winskuo-quic) [reply backends/qualcomm/tests/models.py:467]: We saw `einsum` in  MOEFeedForward under `executorch.examples.models.llama2.llama_transformer`

**Comment by winskuo-quic** [comment]
> This looks great! In the long term, it will be great to seperate changes such that one PR is for one feature. This PR seems including other stuff more than `topK`

Thanks for all the suggestions, and I will separate these into smaller PRs in the future. Sorry that the title is a little misleading. Originally, we were trying to enable `topK` and used `MOEFeedForward` as a test case since there is `topK` in there. However, we ran into other issues like `einsum` decomposition, etc. I will keep the title more accurate in the future and make sure each PR is focusing on one feature.

Thanks!

File: backends/qualcomm/tests/models.py (hunk 3 above)
Comment (cccclai) [reply backends/qualcomm/tests/models.py:467]: Ah I see, so actually the `einsum` is probably not the best for qnn backend, it's better to replace with other way (I try it before) to get a better performance. But it's still good to run `einsum`.

**Comment by cccclai** [comment]
Hey I may need to get https://github.com/pytorch/executorch/pull/5925 checked in first because it needs to get cherry pick. Then I'll merge this one.

**Comment by winskuo-quic** [comment]
> Hey I may need to get #5925 checked in first because it needs to get cherry pick. Then I'll merge this one.

Sounds good! I can also rebase after #5925 is merged if there are any conflicts.

**Comment by cccclai** [comment]
> > Hey I may need to get #5925 checked in first because it needs to get cherry pick. Then I'll merge this one.
> 
> Sounds good! I can also rebase after #5925 is merged if there are any conflicts.

Hi sorry for keeping you waiting, another PR https://github.com/pytorch/executorch/pull/6025 needs to be merged and also needed for beta...the beta release is happening in 2 days and hopefully that is the last PR

**Comment by cccclai** [comment]
Hi it's ready to merge now, can you rebase? Thank you

**Comment by winskuo-quic** [comment]
> Hi it's ready to merge now, can you rebase? Thank you

I have just rebased. Thanks

**Pull request closed without merging**